src/vim9compile.c

```diff
@@ -1865,20 +1865,24 @@ compile_lhs(
 	else if (use_class)
 	{
 	    // for an object or class member get the type of the member
-	    class_T *cl = lhs->lhs_type->tt_class;
+	    class_T	*cl = lhs->lhs_type->tt_class;
+	    omacc_T	access;
+
+	    lhs->lhs_member_type = class_member_type(cl, after + 1,
+					lhs->lhs_end, &lhs->lhs_member_idx,
+					&access);
+	    if (lhs->lhs_member_idx < 0)
+		return FAIL;
+
 	    // If it is private member variable, then accessing it outside the
 	    // class is not allowed.
-	    if (*(after + 1) == '_' && !inside_class(cctx, cl))
+	    if ((access != VIM_ACCESS_ALL) && !inside_class(cctx, cl))
 	    {
 		char_u *m_name = vim_strnsave(after + 1, lhs->lhs_end - after);
 		semsg(_(e_cannot_access_private_member_str), m_name);
 		vim_free(m_name);
 		return FAIL;
 	    }
```

Comment:
I'm just starting to look at this in detail, but when I was working on something similar (which this may supercede) I ended up with something like the following. It doesn't work "as is" but I think conveys the idea.
```
			char *msg = m->ocm_access == VIM_ACCESS_PRIVATE
				      ? e_cannot_access_private_member_str
				      : e_cannot_change_readonly_variable_str;
			semsg(_(msg), m->ocm_name);
			return FAIL;
```

Reply:
This makes sense.  I have updated the PR with this change.